resource-aggregate/cqrs/eventstore/mongodb/getevents.go

```diff
@@ -0,0 +1,186 @@
+package mongodb
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/plgd-dev/cloud/resource-aggregate/cqrs/eventstore"
+	"github.com/plgd-dev/kit/strings"
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo/options"
+)
+
+/// Get array of unique aggregateId values
+func getEventsAggregateIdFilter(groupID string, queries []eventstore.GetEventsQuery) bson.A {
+	if len(queries) == 0 {
+		return nil
+	}
+
+	// get unique aggregateIds
+	aggregateIds := make(map[string]struct{})
+	for _, query := range queries {
+		if query.GroupID == groupID && len(query.AggregateID) != 0 {
+			aggregateIds[query.AggregateID] = struct{}{}
+		}
+	}
+
+	if len(aggregateIds) == 0 {
+		return nil
+	}
+
+	// filter to include only given unique aggregateIds
+	aggregateIdFilter := make(bson.A, 0, len(aggregateIds))
+	for aggregateId := range aggregateIds {
+		aggregateIdFilter = append(aggregateIdFilter, aggregateId)
+	}
+	return aggregateIdFilter
+}
+
+func getEventsFilter(groupID string, queries []eventstore.GetEventsQuery, timestamp int64) bson.D {
+	filter := bson.D{}
+	if len(groupID) > 0 {
+		// filter documents by groupdID
+		filter = append(filter, bson.E{Key: groupIDKey, Value: groupID})
+	}
+
+	if len(queries) > 0 {
+		aggregateIdFilter := getEventsAggregateIdFilter(groupID, queries)
+		if len(aggregateIdFilter) > 0 {
+			// filter documents by aggregateID
+			filter = append(filter, bson.E{
+				Key: aggregateIDKey,
+				Value: bson.M{
+					"$in": aggregateIdFilter,
+				},
+			})
+		}
+	}
+
+	if timestamp > 0 {
+		// filter documents that have the latest timestamp of events larger than given value
+		filter = append(filter, bson.E{
+			Key: latestTimestampKey,
+			Value: bson.M{
+				"$gt": timestamp,
+			},
+		})
+	}
+
+	return filter
+}
+
+func getEventsProjection(timestamp int64) bson.M {
+	projection := bson.M{
+		"_id":          0,
+		groupIDKey:     1,
+		aggregateIDKey: 1,
+		eventsKey:      1,
+	}
+
+	if timestamp > 0 {
+		filter := bson.M{
+			"input": "$" + eventsKey,
+			"as":    "event",
+			"cond": bson.M{
+				"$gt": bson.A{"$$event." + timestampKey, timestamp},
+			},
+		}
+		projection[eventsKey] = bson.M{
+			"$filter": filter,
+		}
+	}
+
+	return projection
+}
+
+func getEventsQueriesToMongoQuery(groupID string, queries []eventstore.GetEventsQuery, timestamp int64) (interface{}, *options.FindOptions) {
+	filter := getEventsFilter(groupID, queries, timestamp)
+
+	opts := options.Find()
```

Comment:
Why you don't use index-es ?
If you add indexes could you compare duration time of test `https://github.com/plgd-dev/cloud/blob/v2/resource-aggregate/cqrs/aggregate/aggregateParralel_test.go#L50` with and without indexes ?

Reply:
Is using `hint` necessary though? Mongodb uses indexes automatically, using hint (https://docs.mongodb.com/manual/reference/operator/meta/hint/) forces the query optimizer to use the selected index. Is the query optimizer slow or why is `hint` necessary?

Reply:
Yes. From my experience if I don't provide hint it uses full scan instead of indexes and find query was very slow.

You can try: - pls create test where you store lot's of events (200000 events) and query half of them via getEvents call and measure the duration. (eg https://github.com/plgd-dev/cloud/blob/v2/resource-aggregate/cqrs/aggregate/aggregateParralel_test.go#L50)